packages/cli/src/utils/api.js

```diff
@@ -197,50 +199,57 @@ const writeLinkedAppConfig = async (app, appDir) => {
   return writeFile(file, prettyJSONstringify(updatedConfig));
 };
 
+const checkCredentials = () => callAPI('/check');
+
 /**
- * gets app details from API
+ * get app info from server and provide specifc error messages if a user doesn't have write permissions for that app
  */
-const getLinkedApp = async appDir => {
-  try {
-    const app = await getLinkedAppConfig(appDir);
-    if (!app) {
-      return {};
-    }
-    return callAPI('/apps/' + app.id);
-  } catch (e) {
-    if (constants.IS_TESTING) {
-      console.error(e);
-    }
+const getWritableApp = async () => {
+  startSpinner('Checking authentication & permissions');
+  await checkCredentials();
+
+  const linkedAppConfig = await getLinkedAppConfig(undefined, false);
+  if (!linkedAppConfig.id) {
     throw new Error(
-      `Unable to complete that operation. Either:
-* your auth file is stale (run \`${colors.cyan('zapier login')}\`)
-* your ${
-        constants.CURRENT_APP_FILE
-      } points to an app you can't access (run \`${colors.cyan(
-        'zapier link'
-      )}\` to refresh the link to an existing app or \`${colors.cyan(
+      `It looks like your integration hasn't been registered. Run \`${colors.cyan(
         'zapier register'
-      )}\` to create a new app).`
+      )}\` before pushing.`
     );
   }
+
+  try {
+    const app = await callAPI(`/apps/${linkedAppConfig.id}`);
+    endSpinner();
+    return app;
+  } catch (err) {
+    // if this fails, we know the issue is they can't see this app
+    const message = `You don't have access to integration ID ${
+      linkedAppConfig.id
+    } (or it doesn't exist). Try running \`${colors.cyan(
+      'zapier link'
+    )}\` to correct that.${
+      process.env.ZAPIER_BASE_ENDPOINT
+        ? "\n\nAlso, it looks like you're testing against a non-production environment. Make sure you've run `zapier link` while providing ZAPIER_BASE_ENDPOINT."
+        : ''
+    }`;
+    throw new Error(message);
```

Comment:
we could! I tried that out and it was a little tougher to reason about what exactly the error was. but! it turns out the server gives different responses for bad auth vs can't see the app. So when running an authed command, you can be in one of the following states:

* no linked app file (`getLinkedAppConfig` returns none, we tell you to `link`)
* no auth file at all (`readCredentials` throws, we tell you to `login`)
* auth file present, but invalid (server returns `401 Not Authenticated`, we tell you to `login`)
* auth file present and valid (you get what you want)

all that with a single API call!

Reply:
Cool, that's even better! Thanks for finding that out! I only have one further suggestion. Instead of checking if the error message contains "Not authenticated" ([line 228](https://github.com/zapier/zapier-platform/pull/216/files#diff-604883b40a4e0331731522508b1d410eR228)), can we check if the response status is 401? I think that's a more reliable interface for the client to rely on.